ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,525 @@
+    .header =
+        {.version = EBPF_HELPER_FUNCTION_ADDRESSES_CURRENT_VERSION,
+         .size = EBPF_HELPER_FUNCTION_ADDRESSES_CURRENT_VERSION_SIZE},
+    .helper_function_count = EBPF_COUNT_OF(_ebpf_netevent_ext_helper_functions),
+    .helper_function_address = (uint64_t*)_ebpf_netevent_ext_helper_functions};
+
+// Context structure for the client module's registration
+typedef struct CLIENT_REGISTRATION_CONTEXT_
+{
+    HANDLE client_registration_handle; // Registration handle (TBD)
+
+} CLIENT_REGISTRATION_CONTEXT, *PCLIENT_REGISTRATION_CONTEXT;
+static CLIENT_REGISTRATION_CONTEXT _netevent_client_registration_context = {.client_registration_handle = NULL};
+
+// Context structure for the client module's binding to a provider module
+typedef struct CLIENT_BINDING_CONTEXT_
+{
+    HANDLE nmr_binding_handle;
+    void* client_context;
+    void* provider_binding_context;
+    const void* provider_dispatch;
+    PNPI_REGISTRATION_INSTANCE provider_registration_instance;
+} CLIENT_BINDING_CONTEXT, *PCLIENT_BINDING_CONTEXT;
+CLIENT_BINDING_CONTEXT _netevent_client_binding_context = {
+    .nmr_binding_handle = NULL,
+    .client_context = NULL,
+    .provider_binding_context = NULL,
+    .provider_dispatch = NULL,
+    .provider_registration_instance = NULL};
+
+// Structure for the extension NMR client module's characteristics
+const NPI_CLIENT_CHARACTERISTICS _netevent_client_characteristics = {
+    0,
+    sizeof(NPI_CLIENT_CHARACTERISTICS),
+    _netevent_ebpf_extension_attach_provider, // Called by NMR after the client module has registered with NMR.
+    _netevent_ebpf_extension_detach_provider,
+    NULL,
+    {0, sizeof(NPI_REGISTRATION_INSTANCE), &netevent_npiid, &netevent_client_module_id, 0, NULL}};
+
+//
+//  NMR client attach/detach callbacks functions to NetEvent as a provider
+//
+NTSTATUS
+_netevent_ebpf_extension_attach_provider(
+    _In_ HANDLE nmr_binding_handle,
+    _In_ void* client_context,
+    _In_ PNPI_REGISTRATION_INSTANCE provider_registration_instance)
+{
+    EBPF_EXT_LOG_ENTRY();
+
+    UNREFERENCED_PARAMETER(client_context);
+    UNREFERENCED_PARAMETER(provider_registration_instance);
+
+    // Attach to the NetEvent provider module.
+    NTSTATUS status = NmrClientAttachProvider(
+        nmr_binding_handle,
+        &_netevent_client_binding_context,
+        &_netevent_client_dispatch,
+        &_netevent_client_binding_context.provider_binding_context,
+        &_netevent_client_binding_context.provider_dispatch);
+    if (!NT_SUCCESS(status)) {
+        EBPF_EXT_LOG_NTSTATUS_API_FAILURE(EBPF_EXT_TRACELOG_KEYWORD_EXTENSION, "NmrRegisterProvider", status);
+        goto Exit;
+    }
+
+Exit:
+    EBPF_EXT_RETURN_NTSTATUS(status);
+}
+
+NTSTATUS
+_netevent_ebpf_extension_detach_provider(_In_ HANDLE nmr_binding_handle)
+{
+    EBPF_EXT_LOG_ENTRY();
+
+    UNREFERENCED_PARAMETER(nmr_binding_handle);
+    // No rundown, since there no state dependency from the provider (i.e. netevent_sim).
```

Comment:
Re. 'No rundown', was curious about how this was verified. A more detailed explanation would be of help, especially considering the potential for races during unload of this extension while in use.

Reply:
This is an out of date comment. Will update.